src/functions/env.ts

```diff
@@ -0,0 +1,238 @@
+    line = line.trim();
+    if (line.startsWith("#")) {
+      // Ignore comments
+      continue;
+    }
+    if (line.length) errors.push(line);
+  }
+
+  return { envs, errors };
+}
+
+/**
+ * Validates string for use as an env var key.
+ *
+ * We restrict key names to ones that conform to POSIX standards.
+ * This is more restrictive than what is allowed in Cloud Functions or Cloud Run.
+ */
+export function validateKey(key: string): void {
+  if (RESERVED_KEYS.includes(key)) {
+    throw new Error("Key reserved for internal use.");
+  }
+  if (!/^[A-Z_][A-Z0-9_]*$/.test(key)) {
+    throw new Error(
+      "Key must start with an uppercase ASCII letter or underscore" +
+        ", and then consist of uppercase ASCII letters, digits, and underscores."
+    );
+  }
+  if (key.startsWith("X_GOOGLE_") || key.startsWith("FIREBASE_")) {
+    throw new Error("Key starts with a reserved prefix (X_GOOGLE_ or FIREBASE_)");
+  }
+}
+
+// Parse dotenv file, but throw errors if:
+//   1. Input has any invalid lines.
+//   2. Any env key is reserved.
+function parseStrict(data: string): Record<string, string> {
+  const { envs, errors } = parse(data);
+
+  if (errors.length) {
+    logger.debug(`Invalid dotenv file, error on lines:\n${errors.join("\n")}`);
+    throw new Error(`Invalid dotenv file, error on lines: ${errors.join(",")}`);
+  }
+
+  const invalidKeys: Record<string, string> = {};
+  for (const key of Object.keys(envs)) {
+    try {
+      validateKey(key);
+    } catch (err) {
+      invalidKeys[key] = err.message || "unknown error";
+    }
+  }
+  if (Object.keys(invalidKeys).length > 0) {
+    logger.debug(
+      `Found invalid keys: ${Object.entries(invalidKeys)
+        .map(([k, v]) => `${k}: ${v}`)
+        .join("\n")}`
+    );
+    throw new Error(`Found invalid keys: ${Object.keys(invalidKeys).join(",")}`);
+  }
+
+  return envs;
+}
+
+/**
+ * Loads environment variables for a project.
+ *
+ * Load looks for .env files at the root of functions source directory
+ * and loads the contents of the .env files.
+ *
+ * .env files are searched and merged in the following order:
+ *
+ *   1. .env
+ *   2. .env.<project or alias> (or .env.local if isEmulator is true).
+ *
+ * If both .env.<project> and .env.<alias> files are found, an error is thrown.
+ *
+ * @return {Record<string, string>} Environment variables for the project.
+ */
+export function load(options: {
+  functionsSource: string;
+  projectId: string;
+  projectAlias?: string;
+  isEmulator?: boolean;
+}): Record<string, string> {
+  const targetFiles = [".env", `.env.${options.projectId}`];
+  if (options.isEmulator) {
+    targetFiles.push(".env.local");
+  } else {
+    targetFiles.push(`.env.${options.projectId}`);
+    if (options.projectAlias && options.projectAlias.length) {
+      targetFiles.push(`.env.${options.projectAlias}`);
+    }
+  }
+
+  const targetPaths = targetFiles
+    .map((f) => path.join(options.functionsSource, f))
+    .filter(fs.existsSync);
```

Comment:
It's much more of a PITA, but I'm wondering if we should be doing this in parallel.

Reply:
Maybeeeeeeeeeee?

I'm for keeping the "slow" implementation since the code read at most 2 files and each file would normally be small, but I also don't know how much more of a PITA it would be to convert this into "parallel" code. If there is some existing implementation I can peek at?

Reply:
On further thought and consulting with the CLI chat room, don't sweat it.